### PR TITLE
Add call frames

### PR DESCRIPTION
**Pull request by fkrause98**
## Changes:
- Added call frames (execution context) for each call.
- Added simple far call implementation.
- Added test that runs a recursive programs, generates frames, and expects a certain amount of spent gas.
- Adapted coded to use the current context.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -90,18 +100,22 @@ pub fn run_program_with_custom_state(bin_path: &str, mut vm: VMState) -> (U256,
...
                     LogOpcode::TransientStorageWrite => todo!(),
                 },
-                Variant::FarCall(_) => todo!(),
+                Variant::FarCall(far_call_variant) => far_call(&mut vm, &far_call_variant),
+                // TODO: This is not how return works. Fix when we have calls between contracts
+                // hooked up.
+                // This is only to keep the context for tests
+                Variant::Ret(_) if vm.running_frames.len() > 1 => {
+                    vm.pop_frame();
+                }
```
Comment (gianbelinche) [inline src/lib.rs:109]: Isn't it better to add this if inside Ret?
```
Variant::Ret(_) => {
    if vm.running_frames.len() > 1 => {
        vm.pop_frame();
    } else {
        break
    }
}
```

File: src/op_handlers/far_call.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+use zkevm_opcode_defs::FarCallOpcode;
+
+use crate::state::VMState;
+
+pub fn far_call(vm: &mut VMState, opcode: &FarCallOpcode) {
+    match opcode {
+        FarCallOpcode::Normal => {
+            let program_code = vm.current_context().code_page.clone();
+            let stipend = vm.current_context().gas_left;
+            vm.push_frame(program_code, stipend.0 / 32);
```
Comment (gianbelinche) [inline src/op_handlers/far_call.rs:10]: This is just a placeholder to test the callframes and not the actual farcall implementation right?

**Review by gianbelinche — CHANGES_REQUESTED** [review]
Remember to fix the Lint

File: src/op_handlers/far_call.rs (hunk 2 above)
Comment (fkrause98) [reply src/op_handlers/far_call.rs:10]: Right, I'll be implementing this on another PR.

File: src/lib.rs (hunk 1 above)
Comment (fkrause98) [reply src/lib.rs:109]: You're right, I'll change it.

File: tests/integration_test.rs [hunk 3]
```diff
@@ -451,26 +452,38 @@ fn test_more_complex_program_with_conditionals() {
...
 }
 #[test]
 // This test should run out of gas before
 // the program can save a number 3 into the storage.
 fn test_runs_out_of_gas_and_stops() {
     let bin_path = make_bin_path_asm("add_with_costs");
-    let vm = VMStateBuilder::new().gas_left(5510).build();
-    let (result, _) = run_program_with_custom_state(&bin_path, vm);
+    let program_code = program_from_file(&bin_path);
+    let frame = CallFrame::new(program_code, 5510);
```
Comment (IAvecilla) [inline tests/integration_test.rs:463]: The idea of creating a custom `CallFrame` here is to manually set up the gas, right? The idea for the future is to always use `load_program(...)` function and let the VM generate the new context.

Comment (fkrause98) [reply tests/integration_test.rs]: Yup, ideally we should always use load_program.

File: src/lib.rs [hunk 4]
```diff
@@ -55,7 +79,9 @@ pub fn run_program(bin_path: &str) -> (U256, VMState) {
 /// Returns the value stored at storage with key 0 and the final vm state.
 pub fn run_program_with_custom_state(bin_path: &str, mut vm: VMState) -> (U256, VMState) {
     let program = program_from_file(bin_path);
-    vm.push_frame(program, DEFAULT_INITIAL_GAS);
+    let storage = RocksDB::open(env::temp_dir()).unwrap();
+    let storage = Rc::new(RefCell::new(storage));
+    vm.push_frame(program, DEFAULT_INITIAL_GAS, storage);
```
Comment (gianbelinche) [inline src/lib.rs:84]: Doesnt this makes `run_program_in_memory` run with rocksdb?

**Review by gianbelinche — CHANGES_REQUESTED** [review]

**Review by gianbelinche — APPROVED** [review]
Great Work!

File: src/lib.rs (hunk 4 above)
Comment (fkrause98) [reply src/lib.rs:88]: fixed

**Review by juan518munoz — APPROVED** [review]
LGTM

**Pull request merged**